resources/testharness.js

```diff
@@ -2920,12 +3096,10 @@ policies and contribution forms [3].
             }
             test.set_status(test.FAIL, e.message, stack);
             test.phase = test.phases.HAS_RESULT;
-            test.done();
```

Comment:
Why no `test.done()` here? Was it always unnecessary, or only now?

Reply:
This is due to a bug in my changes to `test.done`. In the version I originally
submitted, if invoked after a test has already transitioned to the "COMPLETE"
phase, it would exit without calling any "done" callbacks. Effectively, calling
`test.done` here caused the test harness to time out.

I've been thinking about the internal API of this change set. [You previously
identified how the `done_callbacks` abstraction was a bit
leaky](https://github.com/w3c/web-platform-tests/pull/8748#discussion_r158028605), and this is another sign of that.

So I hope you don't mind, but I've refactored the internals for clarity. The
only outward change is that it allows us to re-introduce this `test.done`
invocation (which is superfluous in `master`, by the way--I've included a
comment to remove it later).